File: app/src/main/java/io/github/studerus/pepper_android_realtime/tools/ToolContext.java
package io.github.studerus.pepper_android_realtime.tools;

import android.content.Context;
import com.aldebaran.qi.sdk.QiContext;
import io.github.studerus.pepper_android_realtime.ApiKeyManager;
import io.github.studerus.pepper_android_realtime.ChatActivity;
import io.github.studerus.pepper_android_realtime.MovementController;
import io.github.studerus.pepper_android_realtime.NavigationServiceManager;
import io.github.studerus.pepper_android_realtime.data.LocationProvider;

/**
 * Shared context providing dependencies for all tools.
 * Centralizes access to common services and components.
 */
public class ToolContext {
    
    private final Context appContext;
    private QiContext qiContext; // Made non-final to allow updates
    private final ChatActivity activity;
    private final ApiKeyManager apiKeyManager;
    private final MovementController movementController;
    private final LocationProvider locationProvider;
    
    // Constructor
    public ToolContext(Context appContext, QiContext qiContext, ChatActivity activity,
                      ApiKeyManager apiKeyManager, MovementController movementController,
                      LocationProvider locationProvider) {
        this.appContext = appContext;
        this.qiContext = qiContext;
        this.activity = activity;
        this.apiKeyManager = apiKeyManager;
        this.movementController = movementController;
        this.locationProvider = locationProvider;
    }
    
    // Getters
    public Context getAppContext() { return appContext; }
    public QiContext getQiContext() { return qiContext; }
    public ChatActivity getActivity() { return activity; }
    public ApiKeyManager getApiKeyManager() { return apiKeyManager; }
    @SuppressWarnings("unused")
    public MovementController getMovementController() { return movementController; }
    public LocationProvider getLocationProvider() { return locationProvider; }
    
    /**
     * Get NavigationServiceManager from activity - preferred for navigation operations
     * This coordinates services during navigation unlike direct MovementController access
     */
    public NavigationServiceManager getNavigationServiceManager() {
        return activity != null ? activity.getNavigationServiceManager() : null;
    }
    
    // Convenience methods
    public boolean isQiContextNotReady() { return qiContext == null; }
    public boolean hasUi() { return activity != null; }
    
    /**
     * Update QiContext when robot focus changes
     * @param newQiContext New QiContext or null if focus is lost
     */
    public void updateQiContext(QiContext newQiContext) {
        this.qiContext = newQiContext;
    }
    
    /**
     * Send message directly to Realtime API
     * @param message Message to send
     * @param requestResponse Whether to request a response from AI
     */
    public void sendAsyncUpdate(String message, boolean requestResponse) {
        if (activity != null) {
            activity.sendMessageToRealtimeAPI(message, requestResponse, true);
        }
    }
    
    /**
     * Notify about service state changes for proper service management
     * @param mode Service mode (e.g., "enterLocalizationMode", "resumeNormalOperation")
     */
    public void notifyServiceStateChange(String mode) {
        if (activity != null) {
            activity.handleServiceStateChange(mode);
        }
    }
}

File: app/src/main/java/ch/fhnw/pepper_realtime/tools/ToolContext.kt
package ch.fhnw.pepper_realtime.tools

import android.app.Activity
import android.content.Context
import ch.fhnw.pepper_realtime.controller.GestureController
import ch.fhnw.pepper_realtime.controller.MovementController
import ch.fhnw.pepper_realtime.controller.RobotFocusManager
import ch.fhnw.pepper_realtime.data.LocationProvider
import ch.fhnw.pepper_realtime.manager.ApiKeyManager
import ch.fhnw.pepper_realtime.manager.NavigationServiceManager
import ch.fhnw.pepper_realtime.manager.TouchSensorManager
import ch.fhnw.pepper_realtime.service.PerceptionService
import ch.fhnw.pepper_realtime.tools.interfaces.RealtimeMessageSender
import ch.fhnw.pepper_realtime.tools.interfaces.ToolHost

/**
 * Shared context providing dependencies for all tools.
 * Centralizes access to common services and components.
 * 
 * For Kotlin: Use the properties directly (e.g., context.apiKeyManager)
 * For Java: Use the getXxx() methods (e.g., context.getApiKeyManager())
 */
class ToolContext(
    private val _toolHost: ToolHost,
    private val _robotFocusManager: RobotFocusManager?,
    private val _apiKeyManager: ApiKeyManager,
    private val _movementController: MovementController,
    private val _navigationServiceManager: NavigationServiceManager,
    private val _perceptionService: PerceptionService,
    private val _touchSensorManager: TouchSensorManager,
    private val _gestureController: GestureController,
    private val _locationProvider: LocationProvider,
    private val _messageSender: RealtimeMessageSender?
) {
    /**
     * QiContext for Pepper robot operations
     */
    private var _qiContext: Any? = _robotFocusManager?.qiContext

    // Properties with Java-compatible getters via @get:JvmName
    val apiKeyManager: ApiKeyManager
        get() = _apiKeyManager
    
    val movementController: MovementController
        get() = _movementController
    
    val navigationServiceManager: NavigationServiceManager
        get() = _navigationServiceManager
    
    val perceptionService: PerceptionService
        get() = _perceptionService
    
    val touchSensorManager: TouchSensorManager
        get() = _touchSensorManager
    
    val gestureController: GestureController
        get() = _gestureController
    
    val locationProvider: LocationProvider
        get() = _locationProvider
    
    val toolHost: ToolHost
        get() = _toolHost
    
    val appContext: Context
        get() = _toolHost.getAppContext()
    
    val activity: Activity?
        get() = _toolHost.getActivity()
    
    val qiContext: Any?
        get() = _qiContext
    
    val messageSender: RealtimeMessageSender?
        get() = _messageSender

    /**
     * Check if QiContext is not ready
     */
    fun isQiContextNotReady(): Boolean = _qiContext == null

    /**
     * Check if UI is available
     */
    fun hasUi(): Boolean = _toolHost.getActivity() != null

    /**
     * Update QiContext when robot focus changes
     * @param newQiContext New QiContext (Pepper) or null (Standalone/focus lost)
     */
    fun updateQiContext(newQiContext: Any?) {
        this._qiContext = newQiContext
    }

    /**
     * Send message directly to Realtime API
     * @param message Message to send
     * @param requestResponse Whether to request a response from AI
     */
    fun sendAsyncUpdate(message: String, requestResponse: Boolean) {
        _messageSender?.sendMessageToRealtimeAPI(message, requestResponse, true)
    }

    /**
     * Notify about service state changes for proper service management
     * @param mode Service mode (e.g., "enterLocalizationMode", "resumeNormalOperation")
     */
    fun notifyServiceStateChange(mode: String) {
        _toolHost.handleServiceStateChange(mode)
    }
}

